lib/spack/spack/cmd/diff.py

```diff
@@ -0,0 +1,169 @@
+import spack.cmd.common.arguments as arguments
+import spack.environment as ev
+import spack.util.environment
+
+import llnl.util.tty as tty
+import spack.util.spack_json as sjson
+import operator
+import spack.cmd
+
+description = "compare two specs"
+section = "extensions"
+level = "short"
+
+
+def setup_parser(subparser):
+    """Parser is only constructed so that this prints a nice help
+       message with -h. """
+    arguments.add_common_arguments(
+        subparser, ['specs'])
+
+    subparser.add_argument(
+        '--json',
+        action='store_true',
+        default=False,
+        dest='dump_json',
+        help="Dump json output instead of pretty printing."
+    )
+    subparser.add_argument(
+        '--first',
+        action='store_true',
+        default=False,
+        dest='load_first',
+        help="load the first match if multiple packages match the spec"
+    )
+    subparser.add_argument(
+        '--diff-type',
+        default='all',
+        dest='diff_type',
+        choices=[
+            'all', 'version', 'concrete', 'node', 'node_compiler_set',
+            'node_compiler_version_set', 'node_os_set', 'node_platform_set',
+            'node_target_set', 'variant_set'
+        ],
+        help="select the diff type (defaults to all)"
+    )
+
+
+def bold(string):
+    """Make a header string bold so we can easily see it
+    """
+    return "\033[1m" + string + "\033[0m"
+
+
+def compare_specs(a, b, to_string=False):
+    """Generate a comparison, including diffs (for each side) along with
+    an intersection. We can either print the result to the console, or parse
+    into a json object for the user to save. We return an object that shows
+    the differences, intersection, and names for a pair of specs a and b.
+
+    Arguments:
+        a (spec): the first spec to compare
+        b (spec): the second spec to compare
+        a_name (str): the name of spec a
+        b_name (str): the name of spec b
+        to_string (bool): return an object that can be json dumped
+    """
+    from spack.solver.asp import SpackSolverSetup
+
+    # Prepare a solver setup to parse differences
+    setup = SpackSolverSetup()
+
+    a_facts = set(to_tuple(t) for t in setup.spec_clauses(a))
+    b_facts = set(to_tuple(t) for t in setup.spec_clauses(b))
+
+    # We want to present them to the user as simple key: values
+    intersect = list(a_facts.intersection(b_facts))
+    spec1_not_spec2 = list(a_facts.difference(b_facts))
+    spec2_not_spec1 = list(b_facts.difference(a_facts))
+
+    # We want to show what is the same, and then difference for each
+    return {
+        "intersect": flatten(intersect) if to_string else intersect,
+        "a_not_b": flatten(spec1_not_spec2) if to_string else spec1_not_spec2,
+        "b_not_a": flatten(spec2_not_spec1) if to_string else spec2_not_spec1,
+        "a_name": a.name_version_build_hash,
+        "b_name": b.name_version_build_hash,
+    }
+
+
+def to_tuple(asp_function):
```

Comment:
Instead of this, can you make `AspFunction` in `asp.py` comparable as a tuple?  It should be immutable -- the `__getitem__` method on it is vestigial and can be deleted.  All you need to do is this:

```python
import llnl.util.lang as lang

@lang.key_ordering
class AspFunction(AspObject):
    def _cmp_key(self):
        return (self.name, self.args)
```

And change `self.args = []` to `self.args = ()` in the constructor (it's only ever passed tuples).

With that you should not have to call to_tuple on anything to do set comparisons, this function won't be needed, and we can do cool things with `AspFunctions` elsewhere.

Reply:
cool!

Reply:
Hmm, this didn't seem to work. I did remove the __getitem__ so if it breaks tests I can add it back.

Reply:
We can clean this up later.